sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -873,24 +873,31 @@ object InferFiltersFromGenerate extends Rule[LogicalPlan] {
       if !e.deterministic || e.children.forall(_.foldable) => generate
 
     case generate @ Generate(g, _, false, _, _, _) if canInferFilters(g) =>
-      // Exclude child's constraints to guarantee idempotency
-      val inferredFilters = ExpressionSet(
-        Seq(
-          GreaterThan(Size(g.children.head), Literal(0)),
-          IsNotNull(g.children.head)
-        )
-      ) -- generate.child.constraints
-
-      if (inferredFilters.nonEmpty) {
-        generate.copy(child = Filter(inferredFilters.reduce(And), generate.child))
-      } else {
-        generate
+      assert(g.children.length == 1)
+      g.children.head match {
+        case _: CreateNonNullCollection =>
```

Comment:
Why did you use `CreateNonNullCollection` instead of just checking if `nullable`=false?

Reply:
I thought about using the nullable == false check  something like:

 ```
 private def canInferFilters(g: Generator): Boolean = g match {
    case cg: CollectionGenerator if (cg.children.head.nullable == true) => true
    case _ => false
  }
```

but I think there could still be things we want the size == 0 check to apply to. For instance if doing explode(col2) where col2 contained an empty Map or Array. I think this case you still want the size == 0 check. Please correct me if I'm missing something.